Add ServiceValidationError and translation support (#1961)
home-assistant/developers.home-assistant: docs/internationalization/core.md

| `title`             | Title of the integration.                         |
| `config`            | Translations for the config flow.                 |
| `device_automation` | Translations for device automations.              |

> | `exceptions`        | Translations for for error messages.              |

Reviewer: ```suggestion
| `exceptions`        | Translations for error messages.              |
```

```

### Exceptions

Localization is supported for `HomeAssistantError` and its subclasses.

> The translation strings for exceptions are defined under the `exception` key. An example strings file below describes the different supported keys.

Reviewer: ```suggestion
The translation strings for exceptions are defined under the `exception` key in a `strings.json` file. The example below describes the different supported keys.
```

```json
{
  "exceptions": {
    // Translations for known exceptions
    "invalid_index": {
      "message": "Invalid index selected, expected [0,1,2]. Got {index}"
    }
  }
}

```

> Example of raising exception with localization during a service call:

Reviewer: ```suggestion
Example of raising an exception with localization during a service call:
```

```python
async def async_select_index(hass: HomeAssistant, index: int) -> None:
    """Setup the config entry for my device."""
    try:
        check_index(index)
    except ValueError as ex:
        raise ServiceValidationError(

>             message,

Reviewer 2: @jbouwh where does this variable come from.

Author: Good point. It should be a static string with a message to be logged.

Author: https://github.com/home-assistant/developers.home-assistant/pull/1991
